### PR TITLE
Adds three model regions: scaling, reduce and barrier

### LINKED ISSUE #863: Modify geopmbench so that it can support short running regions
> estimate 5

Introduce a stream region that at startup calibrates its execution by time rather than amount of memory.  This can be executed without threading on a subset of cores to provide a CPU bound short running region with predictable execution time.

Parent story #855

### LINKED ISSUE #922: YAML-backed RawReport cannot read frequency_map agent reports
The yaml-based RawReport reader is unable to read frequency_map agent reports. These reports contain a "Frequency map" section where each entry is prefixed with a '\t' character, instead of spaces. Example:

```
Host: mcfly7
Frequency map:
	0x00000000725e8066:3.7000000000000000e+09
	0x00000000f01e38a3:3.7000000000000000e+09
```

YAML doesn't allow tabs for structural indentation, and the RawReport constructor raises a yaml error:

```
yaml.scanner.ScannerError: while scanning for the next token
found character '\t' that cannot start any token
```

One fix may be to modify the frequency_map agent to prefix these lines with spaces instead of tabs.

### LINKED ISSUE #926: Add a Model region whose completion isn't time-based and uses MPI
> estimate 3

Parent story #918

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Fixes #926 
- Fixes #922
- Fixes #863 
- Scaling region has an integration test.
- Inspection of scaling region results looks reasonable.

### CODE REVIEW

File: src/ReduceModelRegion.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    void ReduceModelRegion::run(void)
+    {
+        int num_rank = 0;
+        int err = MPI_Comm_size(MPI_COMM_WORLD, &num_rank);
+        if (err) {
+            throw Exception("MPI_Comm_size", err, __FILE__, __LINE__);
+        }
+        std::vector<double> send_buffer(m_num_elem, 1.0);
```
Comment (cmcantalupo) [inline src/ReduceModelRegion.cpp]: Allocation in constructor would be better.

File: test_integration/test_scaling_region.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+        """
+        script_dir = os.path.dirname(os.path.realpath(__file__))
+        return os.path.join(script_dir, '.libs', 'test_scaling_region')
+
+    def get_exec_args(self):
+        return []
+
+
+#@util.skip_unless_cpufreq()
```
Comment (cmcantalupo) [inline test_integration/test_scaling_region.py]: Maybe get rid of #?

File: src/ScalingModelRegion.cpp [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            for (size_t it = 0; it != repeat; ++it) {
+                run_atom();
+            }
+            atom_time[trial_idx] = geopm_time_since(&time_0) / repeat;
+        }
+        std::sort(atom_time.begin(), atom_time.end());
+        double median_atom_time = atom_time[median_idx];
+        m_num_atom = big_o_in / median_atom_time;
+        m_num_atom = m_num_atom ? m_num_atom : 1;
+        m_norm = 1.0 / m_num_atom;
```
Comment (dianarg) [inline src/ScalingModelRegion.cpp:140]: where does this get used?

File: test_integration/test_scaling_region.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                            0xba9639ec,
+                            0x61d2be85,
+                            0x28eec3a2,
+                            0xd2b7c6a6,
+                            0x9b8bbb81,
+                            0x40cf3ce8,
+                            0x09f341cf,
+                            0xb1914011,
+                            0xf8ad3d36,
+                            0x07949283]
```
Comment (dianarg) [inline test_integration/test_scaling_region.py:125]: is there some way to discover these from the report?

File: src/ReduceModelRegion.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        std::vector<double> recv_buffer(m_num_elem, 0.0);
```
Comment (dianarg) [inline src/ReduceModelRegion.cpp]: make members

File: test_integration/test_scaling_region.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+    """
+    def write(self):
+        """No configuration files are required.
+
+        """
+        pass
+
+    def get_exec_path(self):
+        """Path to bencmark
```
Comment (dianarg) [inline test_integration/test_scaling_region.py]: benchmark

File: src/ScalingModelRegion.cpp (hunk 3 above)
Comment (dianarg) [reply src/ScalingModelRegion.cpp:140]: used by base class :(

**Review by dianarg — DISMISSED** [review]
Ok as is

File: .gitignore [hunk 7]
```diff
@@ -140,9 +140,11 @@
 /test_integration/*.trace*
 /test_integration/geopm_test_integration
 /test_integration/geopm_test_integration.pyc
+/test_integration/test_ee_scaling_spin_mix
 /test_integration/test_ee_stream_dgemm_mix
 /test_integration/test_omp_outer_loop
 /test_integration/test_profile_policy
+/test_integration/test_scaling_region
```
Comment (bakerbrandond) [inline .gitignore:146]: Missing 'ee' and 'mix'?

File: test_integration/geopm_test_launcher.py [hunk 8]
```diff
@@ -283,12 +283,15 @@ def set_num_cpu(self):
             # Don't use hyper-threads.
             self._num_cpu = cpu_thread_core_socket[2] * cpu_thread_core_socket[3] - 1
 
-    def set_cpu_per_rank(self):
-        try:
-            rank_per_node = int(math.ceil(self._num_rank / self._num_node))
-            self._cpu_per_rank = int(self._num_cpu // rank_per_node)
-        except (AttributeError, TypeError):
-            pass
+    def set_cpu_per_rank(self, cpu_per_rank=None):
```
Comment (bakerbrandond) [inline test_integration/geopm_test_launcher.py:290]: No longer need this changes, right?

File: test_integration/test_ee_scaling_spin_mix.cpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        geopm::Profile &prof = geopm::Profile::default_profile();
+        uint64_t network_spin_rid = prof->region("network_spin", GEOPM_REGION_HINT_NETWORK);
+        std::unique_ptr<geopm::ModelRegion> net_spin_model(geopm::ModelRegion::model_region("spin", net_spin_big_o * mix_idx, is_verbose));
+        for (int rep_idx = 0; rep_idx != repeat; ++rep_idx) {
+            scaling_model->run();
+            prof->enter(network_spin_rid);
+            net_spin_model->run();
+            prof->exit(network_spin_rid);
+        }
+        MPI_Barrier(MPI_COMM_WORLD);
```
Comment (bakerbrandond) [inline test_integration/test_ee_scaling_spin_mix.cpp]: When do we start using the new barrier model region?

File: test_integration/test_ee_scaling_spin_mix.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/env python
```
Comment (bakerbrandond) [inline test_integration/test_ee_scaling_spin_mix.py]: This is just a copy of the steam dgemm mix py file... Needs to be updated with the asserts in my integration test

**Review by bakerbrandond — DISMISSED** [review]

**Review by cmcantalupo — COMMENTED** [review]
done

File: src/ReduceModelRegion.cpp (hunk 5 above)
Comment (cmcantalupo) [reply src/ReduceModelRegion.cpp]: done

File: test_integration/test_scaling_region.py (hunk 6 above)
Comment (cmcantalupo) [reply test_integration/test_scaling_region.py]: done

File: .gitignore (hunk 7 above)
Comment (cmcantalupo) [reply .gitignore:146]: they are in ABC order

File: test_integration/geopm_test_launcher.py (hunk 8 above)
Comment (cmcantalupo) [reply test_integration/geopm_test_launcher.py:290]: We don't need them, but they also don't hurt.  Seems liek a feature we may need later.

File: test_integration/test_ee_scaling_spin_mix.cpp (hunk 9 above)
Comment (cmcantalupo) [reply test_integration/test_ee_scaling_spin_mix.cpp]: its kind of useless unless you are using geopmbench.  Maybe we should delete the BarrierModelRegion.  It is just damaging our test coverage.

File: test_integration/test_ee_scaling_spin_mix.py (hunk 10 above)
Comment (cmcantalupo) [reply test_integration/test_ee_scaling_spin_mix.py]: Removed top commit from this PR

File: test_integration/test_scaling_region.py (hunk 2 above)
Comment (cmcantalupo) [reply test_integration/test_scaling_region.py]: Maybe we will remember later if it fails?

**Comment by cmcantalupo** [comment]
I think this is now good to go.  Hopefully travis will finish soon.

File: test_integration/geopm_test_launcher.py (hunk 8 above)
Comment (bgeltz) [reply test_integration/geopm_test_launcher.py:290]: Some context for what was happening here would be useful.

File: test_integration/test_scaling_region.py (hunk 2 above)
Comment (bgeltz) [reply test_integration/test_scaling_region.py]: This should be removed if it's not necessary.

File: test_integration/test_scaling_region.py (hunk 4 above)
Comment (bgeltz) [reply test_integration/test_scaling_region.py:125]: Why is this array of hashes required?  This is gross.

**Review by bgeltz — CHANGES_REQUESTED** [review]

File: test_integration/geopm_test_launcher.py (hunk 8 above)
Comment (cmcantalupo) [reply test_integration/geopm_test_launcher.py:290]: Would you prefer to delete this patch @bgeltz?  It is not needed for the feature, but without it you cannot control the threads per rank when using the test_launcher.

File: test_integration/test_scaling_region.py (hunk 2 above)
Comment (dianarg) [reply test_integration/test_scaling_region.py]: Can be uncommented; the fix for this decorator is in dev

File: src/ScalingModelRegion.hpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+#include <cstdint>
+#include "ModelRegion.hpp"
+
+namespace geopm
+{
+    class ScalingModelRegion : public ModelRegion
+    {
+        public:
+            ScalingModelRegion(double big_o_in,
+                              int verbosity,
```
Comment (dianarg) [inline src/ScalingModelRegion.hpp]: ws

File: test_integration/geopm_test_launcher.py (hunk 8 above)
Comment (bgeltz) [reply test_integration/geopm_test_launcher.py:290]: I'm fine with it here, I just did not understand what it was fixing.

File: test_integration/test_scaling_region.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+                time = report.get_field(region, 'runtime', 'sec')
+                freq_time.append((freq, time))
+            freq_time.sort()
+            time = list(zip(*freq_time)[1])
+            paired_time = zip(time[0:-1], time[1:])
+            for pt in paired_time:
+                self.assertGreater(pt[0], pt[1])
+
+if __name__ == '__main__':
+    try:
```
Comment (dianarg) [inline test_integration/test_scaling_region.py:189]: remind me why we need this.  Doesn't setUpClass only get called once?

File: test_integration/test_scaling_region.py (hunk 4 above)
Comment (cmcantalupo) [reply test_integration/test_scaling_region.py:125]: We don't have a python interface for geopm_region_hash().  I could run with the monitor just to get a report that has the hashes in it, but that seems worse.  I just run the examples/geopm_hash on the names in the test application. This is test code.

File: test_integration/test_scaling_region.py (hunk 2 above)
Comment (cmcantalupo) [reply test_integration/test_scaling_region.py]: Okay

File: test_integration/test_scaling_region.py (hunk 12 above)
Comment (cmcantalupo) [reply test_integration/test_scaling_region.py:189]: The skip launch is to enable rerun of the tests without executing the benchmark.

**Review by bgeltz — APPROVED** [review]

**Pull request merged**